projects/feature-activation/0001-feature-activation.md

```diff
@@ -0,0 +1,558 @@
+
+The block height when the activation process begins for a certain feature.
+
+#### Activation criteria
+
+Activation criteria, evaluation criteria, or simply criteria, are the conditions necessary to determine if a feature should be activated or not. There are basically two criteria used at this point: block height and miner support through bit signaling. Other criteria could be added in the future.
+
+#### Timeout and timeout height
+
+Timeout height is a block height that represents the end of an activation process for a feature. Timeout is when that height is reached. When timeout is reached, the feature may or may not be activated.
+
+#### Bit signaling
+[Bit signaling]: #bit-signaling
+
+Bit signaling is an activation criteria that may or may not be used in conjunction with block height. This allows miners to send a bit flag in a block, representing that they're ready to support a certain feature. Each bit represents a specific feature. After a certain threshold (percentage) of blocks are found to enable the bit flag for a certain feature during some evaluation interval, that feature is considered active.
+
+This should not be confused with voting. When a feature is defined in the feature activation process, it is assumed that it has been agreed by the community and ideally it should be activated by the end of the process. What bit signaling does, is preventing that some feature is activated before a super majority of miners are ready to support it. It is not voting whether to accept or reject some feature. This was a problem in Bitcoin's SegWit update, where some miners used their influence (through hash rate) to delay the update, for either economical, political, or other obscure reasons. [4][5]
+
+With that being said, it is possible that a feature may or may not be activated at the end of an activation process. To preserve flexibility, that behavior will be configurable on a per-feature basis through the `activate_on_timeout` attribute, explained below. That attribute is analogous to the infamous Bitcoin LOT attribute (lock in on timeout) [6], proposed in BIP 8.
+
+#### Burying feature activations
+[Burying feature activations]: #burying-feature-activations
+
+Burying a feature activation means changing code so "history is erased", and it's not clear that a feature was activated using the activation process.
+
+It is allowed to change code that is conditional on feature activation, but **NOT** to remove a feature from the definition, as described in the [Feature and Criteria definition] section.
+
+## Feature and Criteria definition
+[Feature and Criteria definition]: #feature-and-criteria-definition
+
+The definition of new features available for feature activation should be as simple as defining the feature name and the criteria required to evaluate its activation state. Once defined, ideally, it should not be changed. The `Feature` enum will contain all features, past and future, activated or not:
+
+```python
+class Feature(Enum):
```

Comment:
I feel we should use a separate class for each feature. It would allow each feature to be described and configured in a separate file. Another benefit would be to add custom behavior in the future. What do you think? I also think we should use a separate class per network.

Reply:
Makes sense. Maybe the already specified `Criteria` class serves this purpose, and the change would be just on the `Feature` enum.

Each feature could have its separate file, just defining its `Criteria` instance as a constant. in a `features/nano_contracts.py` file:

```python
NANO_CONTRACTS = Criteria(name='NANO_CONTRACTS', ...)
```

Then, in a `features.py` file, there could be a list of features:

```python
FEATURES: List[Criteria] = [
    NANO_CONTRACTS,
    ...
]
```

Does that work?

Reply:
It does! Thanks!

Reply:
Actually, after revisiting this, I suggest the `FEATURES` list is actually a list inside `HathorSettings`. The individual file with a `Criteria` instance for each feature remains. But the feature list would be defined in the settings. This enforces a specific list is used by the `FeatureService`.

Reply:
Done in https://github.com/HathorNetwork/rfcs/pull/51/commits/7b70d95ec05dc13a16f73f4e6fef31991c960525

Reply:
Thanks!

Reply:
@msbrogli I'm unresolving this discussion, as I'm now implementing this and rethinking it again.

As we're currently implementing settings via yaml in other PRs, I believe each network's yaml file would be the best place to define features for feature activation. So in the yaml we would have something like:

```yaml
# mainnet.yml

features:
  - name: NANO_CONTRACTS
    bit: 0
    start_height: 1000
    threshold: 900
    ...
  - name: MY_OTHER_FEATURE:
    bit: 1
    start_height: 1000
    threshold: 900
    ...
```

Previously, you said:

> I feel we should use a separate class for each feature. It would allow each feature to be described and configured in a separate file.

This solution covers this requirement, as each network has its own separate yaml file.

> Another benefit would be to add custom behavior in the future.

Could you please elaborate on this? What do you mean by custom behavior? Maybe this would be covered by adding another attribute into the feature yaml definiton. Custom behavior would be configured via settings.

> I also think we should use a separate class per network.

I believe everything that is variable by network should be defined in settings, not in code.

Another advantage I see in doing this is that we have a single point where you can see all features defined at once. This helps selecting parameters for a new feature, as you can see for example what bits are being used. What do you think?

Reply:
This is also related to your other comment [here](https://github.com/HathorNetwork/rfcs/commit/7b70d95ec05dc13a16f73f4e6fef31991c960525#r112651756). Yes, each feature definiton works like an "internal object", but is actually just a dictionary in the yaml file. A `NamedTuple` is used to enforce the schema.

I think this case is different than you concern since in the yaml this is simply a dict with scalar values, contrary to an internal object actually being instantiated in python code.

Reply:
Hi @glevco . I initially had some mixed feelings about your suggestion. But, after some thinking, I agree that it is a nice approach. If I got it right, each feature will have its own class (e.g., `NanoContractsFeature`) and a manager would load the classes according to the settings file. Is that right?

On custom behavior, I was thinking that some features might have other methods and attributes that would be useful during the execution of the full node. As an example, the `NanoContractFeature` might have a list of the enabled blueprints and it might also carry some attributes used by the consensus algorithm during the execution of a contract. I mean, these attributes would be part of the feature proposal so they would not be configurable from the settings.

Reply:
> If I got it right, each feature will have its own class (e.g., NanoContractsFeature) and a manager would load the classes according to the settings file. Is that right?

Not exactly. Let's discuss it over a call, I think there's some misunderstanding on the configuration of features

Reply:
As discussed in a meeting, I agree that each feature will be just a boolean flag.